### ISSUE TITLE
🔥 Testing with jest or running `npx expo run` fails with 'Native module RNFBAppModule not found. Re-check module install, linking, configuration, build and install steps.'

### ISSUE THREAD

**Issue by itsramiel** [invertase/react-native-firebase#8189]
# Maintainer Edit:
- **Jest**: For testing with jest, you *must* mock react-native-firebase so it does not attempt to access native modules. We include [an example of this that we use](https://github.com/invertase/react-native-firebase/blob/main/jest.setup.ts), but it is your responsibility
- **Expo** For use of Expo, for example `npx expo run` that won't work. Expo Go doesn't have our react-native-firebase modules. You need to prebuild and then locally compile. We document that https://rnfirebase.io/#local-app-compilation

Original issue follows:

----

## Issue

Testing a react native app with an import from firebase such as `import messaging from '@react-native-firebase/messaging';` causes jest to fail:
```bash
 FAIL  __tests__/App.test.tsx
  ● Test suite failed to run

    Native module RNFBAppModule not found. Re-check module install, linking, configuration, build and install steps.

    > 1 | import messaging from '@react-native-firebase/messaging';
        | ^
      2 | messaging;
      3 |
      4 | function App() {

      at new RNFBNativeEventEmitter (node_modules/@react-native-firebase/app/lib/internal/RNFBNativeEventEmitter.js:25:13)
      at Object.<anonymous> (node_modules/@react-native-firebase/app/lib/internal/RNFBNativeEventEmitter.js:106:16)
      at Object.require (node_modules/@react-native-firebase/app/lib/internal/registry/nativeModule.js:20:1)
      at Object.require (node_modules/@react-native-firebase/app/lib/FirebaseApp.js:18:1)
      at Object.require (node_modules/@react-native-firebase/app/lib/internal/index.js:18:1)
      at Object.require (node_modules/@react-native-firebase/messaging/lib/index.js:28:1)
      at Object.require (App.tsx:1:1)
      at Object.require (__tests__/App.test.tsx:7:1)

Test Suites: 1 failed, 1 total
Tests:       0 total
Snapshots:   0 total
Time:        0.152 s
Ran all test suites.
```
---

### Javascript

<details><summary>Click To Expand</summary>
<p>

#### `package.json`:

```json
{
  "name": "Firebase",
  "version": "0.0.1",
  "private": true,
  "scripts": {
    "android": "react-native run-android",
    "ios": "react-native run-ios",
    "lint": "eslint .",
    "start": "react-native start",
    "test": "jest"
  },
  "dependencies": {
    "@react-native-firebase/app": "^21.6.1",
    "@react-native-firebase/messaging": "^21.6.1",
    "react": "18.3.1",
    "react-native": "0.76.5"
  },
  "devDependencies": {
    "@babel/core": "^7.25.2",
    "@babel/preset-env": "^7.25.3",
    "@babel/runtime": "^7.25.0",
    "@react-native-community/cli": "15.0.1",
    "@react-native-community/cli-platform-android": "15.0.1",
    "@react-native-community/cli-platform-ios": "15.0.1",
    "@react-native/babel-preset": "0.76.5",
    "@react-native/eslint-config": "0.76.5",
    "@react-native/metro-config": "0.76.5",
    "@react-native/typescript-config": "0.76.5",
    "@types/react": "^18.2.6",
    "@types/react-test-renderer": "^18.0.0",
    "babel-jest": "^29.6.3",
    "eslint": "^8.19.0",
    "jest": "^29.6.3",
    "prettier": "2.8.8",
    "react-test-renderer": "18.3.1",
    "typescript": "5.0.4"
  },
  "engines": {
    "node": ">=18"
  },
  "packageManager": "yarn@3.6.4"
}
```

#### `firebase.json` for react-native-firebase v6:

Irrelevant here. Doesnt affect testing environment

```json
# N/A
```

</details>

### iOS

<details><summary>Click To Expand</summary>
<p>

#### `ios/Podfile`:

Irrelevant

- [ ] I'm not using Pods
- [x] I'm using Pods and my Podfile looks like:

```ruby
# N/A
```

#### `AppDelegate.m`:

Irrelevant

```objc
// N/A
```

</p>
</details>

---

### Android

<details><summary>Click To Expand</summary>
<p>

#### Have you converted to AndroidX?

Irrelevant

- [ ] my application is an AndroidX application?
- [ ] I am using `android/gradle.settings` `jetifier=true` for Android compatibility?
- [ ] I am using the NPM package `jetifier` for react-native compatibility?

#### `android/build.gradle`:

```groovy
// N/A
```

#### `android/app/build.gradle`:

```groovy
// N/A
```

#### `android/settings.gradle`:

```groovy
// N/A
```

#### `MainApplication.java`:

```java
// N/A
```

#### `AndroidManifest.xml`:

```xml
<!-- N/A -->
```

</p>
</details>

---

## Environment

<details><summary>Click To Expand</summary>
<p>

**`react-native info` output:**

```
info Fetching system and libraries information...
System:
  OS: macOS 15.1
  CPU: (14) arm64 Apple M3 Max
  Memory: 1.75 GB / 36.00 GB
  Shell:
    version: "5.9"
    path: /bin/zsh
Binaries:
  Node:
    version: 20.11.1
    path: /private/var/folders/p1/47x14k7x3hs1y75l4l8qd9980000gn/T/xfs-2ded1f97/node
  Yarn:
    version: 3.6.4
    path: /private/var/folders/p1/47x14k7x3hs1y75l4l8qd9980000gn/T/xfs-2ded1f97/yarn
  npm:
    version: 10.2.4
    path: ~/.nvm/versions/node/v20.11.1/bin/npm
  Watchman:
    version: 2024.10.28.00
    path: /opt/homebrew/bin/watchman
Managers:
  CocoaPods:
    version: 1.14.3
    path: /Users/itsramiel/.rbenv/shims/pod
SDKs:
  iOS SDK:
    Platforms:
      - DriverKit 24.1
      - iOS 18.1
      - macOS 15.1
      - tvOS 18.1
      - visionOS 2.1
      - watchOS 11.1
  Android SDK: Not Found
IDEs:
  Android Studio: 2024.1 AI-241.18034.62.2412.12266719
  Xcode:
    version: 16.1/16B40
    path: /usr/bin/xcodebuild
Languages:
  Java:
    version: 17.0.10
    path: /usr/bin/javac
  Ruby:
    version: 3.0.0
    path: /Users/itsramiel/.rbenv/shims/ruby
npmPackages:
  "@react-native-community/cli":
    installed: 15.0.1
    wanted: 15.0.1
  react:
    installed: 18.3.1
    wanted: 18.3.1
  react-native:
    installed: 0.76.5
    wanted: 0.76.5
  react-native-macos: Not Found
npmGlobalPackages:
  "*react-native*": Not Found
Android:
  hermesEnabled: true
  newArchEnabled: true
iOS:
  hermesEnabled: Not found
  newArchEnabled: false
```

- **Platform that you're experiencing the issue on**: None(testing env, jest)
  - [ ] iOS
  - [ ] Android
  - [ ] **iOS** but have not tested behavior on Android
  - [ ] **Android** but have not tested behavior on iOS
  - [ ] Both
- **`react-native-firebase` version you're using that has this issue:**
  - `e.g. 5.4.3`
- **`Firebase` module(s) you're using that has the issue:**
  - `e.g. Instance ID`
- **Are you using `TypeScript`?**
  - `Y/N` & `VERSION`

</p>
</details>

---

- 👉 Check out [`React Native Firebase`](https://twitter.com/rnfirebase) and [`Invertase`](https://twitter.com/invertaseio) on Twitter for updates on the library.

### Info I find relevant:

So basically you can reproduce with:
```ts
import messaging from '@react-native-firebase/messaging';
messaging;

function App() {
  return null;
}

export default App;
```

The error is this:
```
    Native module RNFBAppModule not found. Re-check module install, linking, configuration, build and install steps.

    > 1 | import messaging from '@react-native-firebase/messaging';
        | ^
      2 | messaging;
      3 |
      4 | function App() {

      at new RNFBNativeEventEmitter (node_modules/@react-native-firebase/app/lib/internal/RNFBNativeEventEmitter.js:25:13)
      at Object.<anonymous> (node_modules/@react-native-firebase/app/lib/internal/RNFBNativeEventEmitter.js:106:16)
      at Object.require (node_modules/@react-native-firebase/app/lib/internal/registry/nativeModule.js:20:1)
      at Object.require (node_modules/@react-native-firebase/app/lib/FirebaseApp.js:18:1)
      at Object.require (node_modules/@react-native-firebase/app/lib/internal/index.js:18:1)
      at Object.require (node_modules/@react-native-firebase/messaging/lib/index.js:28:1)
      at Object.require (App.tsx:1:1)
      at Object.require (__tests__/App.test.tsx:7:1)
 ```
 
 The problem is that this import:
 `      at Object.require (node_modules/@react-native-firebase/app/lib/internal/registry/nativeModule.js:20:1)`
 
 is resolving to `app/lib/internal/nativeModule.ios.js` since apparently ios is the default platform in react native preset for jest.
 
 Now this file imports from: app/lib/internal/nativeModuleAndroidIos.js
 
 which tries to get the module from `NativeModules` which obv will return undefined in a jest env:
 https://github.com/invertase/react-native-firebase/blob/3623a504a6d07fa148390bd3f97d6c06fea8955d/packages/app/lib/internal/nativeModuleAndroidIos.js#L11
 
 which leads to the throwing of the error:
 
 https://github.com/invertase/react-native-firebase/blob/3623a504a6d07fa148390bd3f97d6c06fea8955d/packages/app/lib/internal/RNFBNativeEventEmitter.js#L23-L28

**Comment by mikehardy** [comment]
Using jest to test modules that requires native modules will require mocking and setup of course, this is a typical jest use case you have and a typical jest need for that use case

A follow-on in this instance should be: well, how do they do it then? That is, how do we do our testing in our repo so it works for us?

https://github.com/invertase/react-native-firebase/blob/main/jest.config.js

which has

https://github.com/invertase/react-native-firebase/blob/3623a504a6d07fa148390bd3f97d6c06fea8955d/jest.config.js#L8

Which points to

https://github.com/invertase/react-native-firebase/blob/main/jest.setup.ts

I'd do what we do in that file - or! if there is a better way - please suggest anything better for that file as we use jest too and we welcome PRs. 

Should work?

**Comment by itsramiel** [comment]
What I would love to have is a mock file I could import and then mock the libary with that mock. This is a common thing that other libraries do, examples:
```js
import mockRNCNetInfo from '@react-native-community/netinfo/jest/netinfo-mock.js';
jest.mock('@react-native-community/netinfo', () => mockRNCNetInfo);
```
```js
import mockRNLocalize from 'react-native-localize/mock/index.js';
jest.mock('react-native-localize', () => mockRNLocalize);
```
```js
jest.mock('@react-native-async-storage/async-storage', () =>
  require('@react-native-async-storage/async-storage/jest/async-storage-mock')
);
```

If the library would provide such a mock that can be used internally plus for library users, then that would be perfect.

Note: The failing of jest is new. This didnt use to happen in `18.7.1` but did happen when upgrading to `21.6.1`. [Same was experiences by others](https://stackoverflow.com/questions/79177592/error-native-module-rnfbappmodule-not-found-re-check-module-install-linking), but their solution was to downgrade which is obv not the correct solution.

**Comment by itsramiel** [comment]
Yes, copy pasting `jest.setup.js` into my projects `jest.setup.js` does fix the tests, but the file is mocking `react-native` not the library. Shouldnt we expose a mock of the library instead? Or how is this setup file meant to be used? I am not sure if exposing a setup file that mocks react native is the best here

**Comment by mikehardy** [comment]
It mocks react-native yes, because we do module loading. It mocks the minimum to make that "work" (read as: not fail) in a jest environment, then for the rest of react-native, the mock is the actual implementation of react-native

What could possibly go wrong ;-)

You tested it and it worked for you, that's great news. No reason we couldn't ship it and people could use it if they like

**Comment by itsramiel** [comment]
> What could possibly go wrong ;-)

I am not familiar with jest but my initial thought was that if we mock rn, then does that override other mocks of rn that a user may already have. I am not very familiar with jest. maybe you already know, or I have to test

**Comment by mikehardy** [comment]
You'd have to test that, however if jest is already mocking things then I would assume (dangerous!) that when *our* jest says "import" for react-native we'll get the mocked one.

it may be then that our mocks will collide with other specific mocks, but ours is mocking the minimum we need - not sure what we can do about that.

Then the rest of the react native module from import is spread into our mock so it should still apply

I'm not sure a different way to do it though, we have a multi-module system and need to reach into react-native and inspect / load etc our native modules. In a jest context that has to be mocked.

**Comment by GideonAgboba** [comment]
THIS ERROR OCCURS IN BRIDGELESS MODE. PLEASE CAN WE UPGRADE THIS LIBRARY TO SUPPORT BRIDGELESS MODE?

**Comment by mikehardy** [comment]
@GideonAgboba I DON'T BELIEVE JEST CAN TELL WHETHER SOMETHING IS BRIDGELESS OR NOT, YOU HAVE TO MOCK THINGS EITHER WAY

**Comment by danielctchin** [comment]
Hi there anyone got this to work? Had the same issue with npx expo start, however on npx expo run, it works.
Error: Native module RNFBAppModule not found. Re-check module install, linking, configuration, build and install steps.

**Comment by mikehardy** [comment]
> on npx expo run, it works

Sounds like you go it to run?

> Error: Native module RNFBAppModule not found. Re-check module install, linking, configuration, build and install steps

"Re-check .... build ... steps"

When using Expo with custom native modules getting those things exactly right is critical

**Comment by danielctchin** [comment]
Thanks @mikehardy. 
I got it to work if its expo run.
But on development environment, we usually use npx expo start. And by running this comment we still hit the error above.
Do we need to import some files?

**Comment by mikehardy** [comment]
Apologies @danielctchin I don't use Expo so I don't know. If `expo run` works, use that is what I'd do

**Comment by tim5go** [comment]
Hi @mikehardy 
I'm facing the same issue with `import analytics from '@react-native-firebase/analytics;`.

For developers who use Expo, getting `npx expo start` to work is crucial as it allows for real-time code changes without needing a full rebuild. I hope you understand our difficulty with this. Thank you!

**Comment by mikehardy** [comment]
How did expo run work for you? Do you have any other suggestions? You'll have to understand I don't use expo so I'm not able to troubleshoot this

**Comment by mikehardy** [comment]
I believe that was part of getting web to work as an underlying fallback and to implement "other" platform support (e.g., macOS)

Here's a different question though, taking the opposite perspective:

How could react-native-firebase work at all, if it is unable to locate the native app module? (answer: it cannot, at least *not on a native platform*)

Follow-on question: is it better to determine quickly and fail quickly if the underlying app module is not found, or to quietly proceed and then fail later? (answer: fail fast is typically better, allowing for exceptions)

So what we have here is a change that attempts to locate the native module immediately and fails if it cannot. Which...seems like a good change. At least for the native platforms

Why does `npx expo start` not correctly install + run the *native* modules?

 Is `npx expo start` a purely web command, with no expectation of the native modules being available? If so, and there is the expectation that react-native-firebase works perfectly in a web-only context, I'm not sure that's a correct expectation at the moment. react-native-firebase still at its heart relies on native modules, with some firebase-js-sdk backfill for modules that aren't available on "other" platforms. We hope to be 100% web compatible in the future but we're not there yet I don't think, and still require "shims" to work on web-only.

Indeed, `npx expo start` appears to rely on the Expo Go app which will *never* work for react-native-firebase because it does not include the built-in react-native-firebase modules, nor will it ever include them.

So perhaps there is some variation of the expo cli stuff you need to use instead, like 

https://docs.expo.dev/more/expo-cli/#prebuild and
https://docs.expo.dev/more/expo-cli/#compiling-android

**Comment by mikehardy** [comment]
Also worth noting:
- the original poster is talking about *jest*, which is more like a web context, and requires jest mocking
- this issue has been mostly hijacked and is now "hey, `npx expo start` doesn't work, which is somewhat related to the original issue but is really a separate issue
- we have documentation that explains you need to use prebuild etc: https://rnfirebase.io/#local-app-compilation

**Comment by mikehardy** [comment]
The stackoverflow link was not related to jest or expo and so was also not relevant, however it's got all the juicy keywords to show up in searches for a handful of related searches - just as this issue does. So it acts as a magnet perhaps for people. I responded there as well, attempting to answer both this question and the original posters question on SO https://stackoverflow.com/questions/79177592/error-native-module-rnfbappmodule-not-found-re-check-module-install-linking/79378252#79378252

**Comment by tim5go** [comment]
@mikehardy 
Your explanation is very clear, thanks  🙏🏻

**Comment by mikehardy** [comment]
I have edited the original description to directly include how to use jest or expo here, and updated the title.
I believe both cases are resolved so I'm closing this issue but it isn't locked or anything.
If there's something missing conversation may of course continue and we'll all try to help each other
Cheers

**Comment by sberkaye** [comment]
@mikehardy I've been using react-native-firebase on Expo after following the steps on the documentation, creating a clean prebuild and then compiling locally. ([as documented here](https://docs.expo.dev/guides/local-app-development/#local-app-compilation)) I had no issues running it on iOS, but on Android it gives the same "Error: Native module RNFBAppModule not found. Re-check module install, linking, configuration, build and install steps." error after compiling & running. I had deleted `android` folder to create clean prebuilds, ran `npx expo run prebuild --clean` and then built with `npx expo run:android` but it didn't work. I saw your comment here & on StackOverflow but unfortunately following the documentation didn't work for me. Running `npx expo run:ios` works completely fine though. Here are my package versions:

    `"@react-native-firebase/app": "^21.7.1",
    "@react-native-firebase/auth": "^21.7.1",
    "expo": "~52.0.11",
    "react-native": "0.76.3",`


Can you provide any insights? Thanks in advance!

**Comment by nicolasguzca** [comment]
On the latest version of Expo and the library you can fix it by excluding the native dependencies like` @react-native-firebase/app `from development builds (Expo Go) and only use them in production/compiled builds.

I basically just created a   mock for Firebase in development mode

```
import Constants from "expo-constants";

let analytics: any;

// Mock analytics in Expo Go (development)
if (Constants.appOwnership === "expo") {
  console.log("Running in development - Firebase Analytics is mocked.");

  analytics = {
    logEvent: (event: string, params?: object) =>
      console.log(`Mock logEvent called: ${event}`, params),
    setUserId: (id: string) => console.log(`Mock setUserId called with: ${id}`),
    setUserProperties: (props: object) =>
      console.log(`Mock setUserProperties:`, props),
  };
} else {
  // Dynamically import Firebase analytics for production builds only
  analytics = require("@react-native-firebase/analytics").default;
}

export default analytics;

```

And then replace` import analytics from "@react-native-firebase/analytics";` with `import { firebase } from './firebaseConfig';` (use the path specific to your project.

**Comment by mikehardy** [comment]
Must admit I'm shocked at the idea of developing with key features of the app disabled during the time when you're supposed to be developing key features, but I suppose that's because I'm used to using firestore and messaging and they're integral parts of the app.

If you are able to successfully develop with firebase disabled during most of your development cycle, wouldn't firebase-js-sdk make more sense? Honestly curious.

**Comment by nicolasguzca** [comment]
Agreed, it depends largely on the task at hand. Expo Go is convenient when you need to quickly check UI changes or test API integrations, thanks to its significantly lower build time. It’s been working well for us in that regard. If not then Development Builds,

For this project, I’m primarily using analytics, so disabling it on Expo Go is fine for now. Hopefully, Expo will update its packages in the future to support native integration with React Native Firebase.

I’m also using firebase-js-sdk for Firestore in some small features, although it occasionally throws console warnings and errors due to its attempts to use web modules.

**Comment by mikehardy** [comment]
> Hopefully, Expo will update its packages in the future to support native integration with React Native Firebase.

I'd be surprised if they ever did that - adding Firebase would probably double the size of the app, and while as a react-native module react-native-firebase appears to be used in maybe 20%+ of apps, that's a lot of suffer to put the other 80% through.

I can see how a shim might make sense in this use case though - if firebase usage is really light but you still definitely want the native SDK powers for some things (like some of the auto-tracked analytics events...) it can make sense. Thanks for sharing

**Comment by raxng** [comment]
For Expo users...

After many hours of trying to get this to work, this recent solution posted on stack overflow worked for me (https://stackoverflow.com/a/79423321/29566034)

Reposted here for convenience and preservation:
<img width="935" alt="Image" src="https://github.com/user-attachments/assets/d490d309-55d9-4d2f-a6db-40374da8bb2a" />

NOTE: The step that stumped me for hours was not doing the following step: **"uninstalling the app from the simulator"**.  For some reason, it was not intuitive for me to do this (I'm a bit of an Expo noob)

I rebuilt everything from scratch, cleared caches, restarted the simulator, etc.  A couple of times the Expo app was re-installed on the simulator overwriting what was already there.  The one thing I didn't try was to completely remove the Expo app from iOS simulator... once I did that, everything worked.

Environment 
- Node v22.13.1
- react-native-firebase v21.7.x
- iPhone 15 iOS Simulator
- Expo development build (not Expo Go)

I hope this helps others... good luck!

**Comment by mikehardy** [comment]
@raxng nice one. I clicked the "Edit" button top right of all docs pages on rnfirebase.io and added this paragraph to the Expo section, seems like it could help others https://github.com/invertase/react-native-firebase/pull/8305

Any one else can do the same - as you discover things that may need updating or further explanation for the Expo-based workflow, please hit that edit button, the Github web UI for docs PRs only takes a minute or two
